### PR TITLE
Fix XAML Visibility for `display:none`

### LINKED ISSUE #6355: Components under "display: none" still participate in focus, UIA, etc
RNW will pass the "display: none" property to Yoga for layout, but doesn't take it into consideration for XAML visibility. This means we create the same XAML tree as normal, but elements are hidden by virtue of being sized as (0,0).

Because of this, components under "display: none" still participate in UIA, focus, or other bits that they likely shouldn't. @kmelmon mentioned this also defeats certain XAML perf optimizations. Concretely, this causes issues with RNTester refactoring, leading to multiple UIA elements with the same automation ID to be displayed at once.

We should propagate this to the XAML tree (maybe just mark the tree as collapsed?).

### PR DESCRIPTION
**Pull request by AgneLukoseviciute**
As mentioned in #6355, components with display:none still participate in focus and UIA. This adds new logic in FrameworkElementViewManager to change visbility to collapsed when display: is set to none. 

I tested the behavior through a new example in RNTester View component page using TextInput to confirm that focus is disabled when display:none. 

Fixes #6355, also likely fixes react-navigation [#9109](https://github.com/react-navigation/react-navigation/issues/9109).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7363)

### CODE REVIEW

**Comment by AgneLukoseviciute** [comment]
Question: from this snapshot and inspecting component properties from VS, it seems like the visibility remains as "visibile," despite the behavior changing as if it was collapsing(and from setting breakpoints in FrameworkElementViewManager I can see that the right code is getting hit and the Visibility value gets set to "Collapsed" - but I never see this in XAML) - any idea what's going on?

**Comment by asklar** [comment]
> Question: from this snapshot and inspecting component properties from VS, it seems like the visibility remains as "visibile," despite the behavior changing as if it was collapsing(and from setting breakpoints in FrameworkElementViewManager I can see that the right code is getting hit and the Visibility value gets set to "Collapsed" - but I never see this in XAML) - any idea what's going on?

if the tree dump says it's not collapsed then it's probably not collapsed : ) (and something else may be re-setting the property?)
Do you see the collapsed element in the UI? What does the Live Visual Tree and Property panes say?

File: packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js [hunk 1]
```diff
@@ -517,6 +519,71 @@ exports.examples = [
...
+  type State = {|
+    displayNone: boolean,
+      displayNone2: boolean,
+        text: string,
+          text2: string,
+      |};
+
+class DisplayNoneVisibilityExample extends React.Component<Props, State> {
+  state = {
+    displayNone: false,
```
Comment (rectified95) [inline packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js:535]: nit: you could also just store the string value here, instead of the boolean

File: packages/e2e-test-app/test/ViewComponentTest.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+describe('ViewComponentTest', () => {
+    /* Test case #1: display:none disabled */
+    test('ViewComponentTestWithoutDisplayNone', async () => {
+        const dump = await dumpVisualTree('view-component-switch-view');
+        expect(dump).toMatchSnapshot();
+    });
+
+    /* Test case #2: Click button once, update view*/
+    test('ViewComponentTestWithDisplayNone', async () => {
+        await toggleDisplayNone();
```
Comment (rectified95) [inline packages/e2e-test-app/test/ViewComponentTest.test.ts:23]: where is this function defined?

File: packages/e2e-test-app/test/ViewComponentTest.test.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        const dump = await dumpVisualTree('view-component-switch-view');
+        expect(dump).toMatchSnapshot();
+    });
+
+    /* Test case #3: Click button one more, return to #1*/
```
Comment (rectified95) [inline packages/e2e-test-app/test/ViewComponentTest.test.ts:28]: don't think we need this case - it basically tests if the Switch works. the first 2 tests cover both cases of visibility already. Also, each of these is a separate test, so you wouldn't 'return' to a previous state because there is none

**Comment by rectified95** [comment]
Overall looks good. Just make sure you figure out the xaml dump issue

File: change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (asklar) [inline change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json:1]: [putting the convo here so we get threading]

> > > Question: from this snapshot and inspecting component properties from VS, it seems like the visibility remains as "visibile," despite the behavior changing as if it was collapsing(and from setting breakpoints in FrameworkElementViewManager I can see that the right code is getting hit and the Visibility value gets set to "Collapsed" - but I never see this in XAML) - any idea what's going on?
> >
> > if the tree dump says it's not collapsed then it's probably not collapsed : ) (and something else may be re-setting the property?)
> Do you see the collapsed element in the UI? What does the Live Visual Tree and Property panes say?
> 
> The "collapsed" element is no longer seen in the UI, but the Live Visual Tree and accompanying properties do not seem to change when display:none is toggled. If I undo the code change I added, same thing happens - component goes out of view when display is set to none, but you are still able to tab into those components - when I add the change back in, I'm no longer able to tab into components which are set to display:none, but according to the properties the visibility is still "visible"...

Comment (rectified95) [reply change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json:1]: Sometimes you need to refresh the live tree view - does it remain collapsed after you 'f5' in VS and refresh the inspector?

Comment (asklar) [reply change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json:1]: There is currently a bug in the live visual tree/property pane where it won't refresh: https://developercommunity.visualstudio.com/t/Live-visual-tree-properties-do-not-refre/1359952

Given that, the only trustworthy option right now seems to be to look at the xaml tree in WinDBG -  we have a debugger extension that lets you do that, we can follow up offline

**Comment by rectified95** [comment]
Also, the JS checks are failing. You need to run `yarn lint:fix` - it'll resolve formatting issues. Also run `yarn flow-check` (with/without '-', dont remember)

**Comment by AgneLukoseviciute** [comment]
> Also, the JS checks are failing. You need to run `yarn lint:fix` - it'll resolve formatting issues. Also run `yarn flow-check` (with/without '-', dont remember)

Will do, I ran `yarn lint:fix` but missed the latter.

File: vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp [hunk 5]
```diff
@@ -535,6 +535,15 @@ bool FrameworkElementViewManager::UpdateProperty(
     } else if (propertyName == "accessibilityActions") {
       auto value = json_type_traits<winrt::IVector<winrt::react::uwp::AccessibilityAction>>::parseJson(propertyValue);
       DynamicAutomationProperties::SetAccessibilityActions(element, value);
+    } else if (propertyName == "display") {
+      if (propertyValue.Type() == winrt::Microsoft::ReactNative::JSValueType::String) {
+        auto value = propertyValue.AsString();
+        if (value == "none") {
+          element.Visibility(winrt::Windows::UI::Xaml::Visibility::Collapsed);
```
Comment (rectified95) [inline vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:542]: Check out our namespace aliases - the winrt:ui:xaml bit should be `xaml` 
Look in CPPWinrtIncludes.h@16

Comment (rectified95) [reply vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:542]: It's currently failing the WINUI3 build in the CI

**Comment by rectified95** [comment]
I've looked i the CI logs, and there are ecessively white spaces - you must have run lint:fix earlier but it didn't get committed.

File: packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js (hunk 1 above)
Comment (AgneLukoseviciute) [reply packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js:535]: I switched to string for a moment, but then I'd have to make the switch press handlers messier :' ( I think I might just keep this if that's good.

File: change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json (hunk 4 above)
Comment (AgneLukoseviciute) [reply change/react-native-windows-66f4660e-05be-46b7-8321-74ea94fb2b4b.json:1]: Confirmed that's it's collapsing 👍 but filed #7392 as a separate issue for testing.. removing snapshots for now, I'll add those back in once that gets fixed.

**Comment by asklar** [comment]
@AgneLukoseviciute looks like the test was removed, was that intentional?

**Comment by AgneLukoseviciute** [comment]
> @AgneLukoseviciute looks like the test was removed, was that intentional?

Yep, I filed an issue(#7392) that will need to get fixed before I could get the tests working. The way the tests search in the e2e app isn't filtering the results down enough, at least in the case of View.

**Comment by asklar** [comment]
Thanks @AgneLukoseviciute ; in the meantime, can we add the test to a different/new page instead?

**Comment by AgneLukoseviciute** [comment]
> Thanks @AgneLukoseviciute ; in the meantime, can we add the test to a different/new page instead?

Yep that should work - I'll add it.

File: packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js [hunk 6]
```diff
@@ -30,6 +30,11 @@ const ComponentExamples: Array<RNTesterExample> = [
     category: 'UI',
     module: require('../examples-win/DatePicker/DatePickerExample'),
   },
+  {
+    key: 'Display:NoneExample',
+    category: 'UI',
+    module: require('../examples-win/DisplayNone/DisplayNoneExample'),
```
Comment (asklar) [inline packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js:36]: I think you might have missed adding the file?

Comment (AgneLukoseviciute) [reply packages/@react-native-windows/tester/src/js/utils/RNTesterList.windows.js:36]: yep, struggling with .gitignore today.. should be included now. About to add the snapshot testing back in to target the new page.

File: packages/@react-native-windows/tester/js/examples-win/DisplayNone/.vs/ProjectSettings.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": null
```
Comment (asklar) [inline packages/@react-native-windows/tester/js/examples-win/DisplayNone/.vs/ProjectSettings.json:2]: don't add files under the .vs folder, those are autogenerated by Visual Studio (includes things like local user settings, intellisense cache, etc.)

File: packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.d.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ * @format
+ */
+/// <reference types="react" />
```
Comment (asklar) [inline packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.d.ts:6]: You don't need to check these in (.d.ts, .windows.js, .windows.js.map), only the TSX file, and the build would produce the others, so you can remove these from the PR

File: packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js [hunk 9]
```diff
@@ -17,8 +17,6 @@ const {
   Text,
   TouchableWithoutFeedback,
   View,
-  Switch,
```
Comment (asklar) [inline packages/@react-native-windows/tester/src/js/examples/View/ViewExample.windows.js]: this file should be deleted/not added too

**Review by asklar — APPROVED** [review]

File: packages/@react-native-windows/tester/src/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+  _handlePress = () => {
+    this.setState({displayNone: !this.state.displayNone});
+  };
+
+  _handlePress2 = () => {
+    this.setState({displayNone2: !this.state.displayNone2});
+  };
+
+  _handleChangeText = text => {
```
Comment (asklar) [inline packages/@react-native-windows/tester/src/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx:66]: looks like you're missing the type on `text` here and below

**Comment by asklar** [comment]
@AgneLukoseviciute  Looks like your snapshots are missing the Left/Top attached properties which the tests recently started requiring. Can you rebase your changes onto the latest master and re-generate the snapshot?

File: packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx [hunk 11]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (rectified95) [inline packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx:1]: this file shouldn't be checked in - should only include files under the /src directory. This one is the result of running yarn build which copies files to where metro pickes it up from

**Review by rectified95 — CHANGES_REQUESTED** [review]
Remember to remove the duplicate file

File: packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx (hunk 11 above)
Comment (asklar) [reply packages/@react-native-windows/tester/js/examples-win/DisplayNone/DisplayNoneExample.windows.tsx:1]: good catch; the tsx extension fooled me. Is this file needed in the compiled output or only the .js?

**Comment by AgneLukoseviciute** [comment]
> Remember to remove the duplicate file

Done, thanks for the catch. From rebasing to latest master all of the commits were added to this PR, is this to be expected? or is there a better way to rebase in the future?

**Comment by asklar** [comment]
> > Remember to remove the duplicate file
> 
> Done, thanks for the catch. From rebasing to latest master all of the commits were added to this PR, is this to be expected? or is there a better way to rebase in the future?

That'll happen when you do a merge instead of a rebase. Sync offline w/Igor or me to fix that.

**Review by rectified95 — APPROVED** [review]

File: vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp [hunk 12]
```diff
@@ -535,6 +535,15 @@ bool FrameworkElementViewManager::UpdateProperty(
     } else if (propertyName == "accessibilityActions") {
       auto value = json_type_traits<winrt::IVector<winrt::react::uwp::AccessibilityAction>>::parseJson(propertyValue);
       DynamicAutomationProperties::SetAccessibilityActions(element, value);
+    } else if (propertyName == "display") {
+      if (propertyValue.Type() == winrt::Microsoft::ReactNative::JSValueType::String) {
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:539]: I think you'll also want to handle propertyValue.IsNull().

Comment (AgneLukoseviciute) [reply vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:539]: I think in that case we'd just want to default to visible, added that in :)

**Review by asklar — APPROVED** [review]

File: vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp (hunk 12 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:539]: I think you probably want to reset to default by calling ClearProperty(UIElement::VisibilityProperty()), but setting explicitly also works.

Comment (AgneLukoseviciute) [reply vnext/Microsoft.ReactNative/Views/FrameworkElementViewManager.cpp:539]: I like that, that matches the pattern for other properties. Thanks!

**Comment by AgneLukoseviciute** [comment]
> > > Question: from this snapshot and inspecting component properties from VS, it seems like the visibility remains as "visibile," despite the behavior changing as if it was collapsing(and from setting breakpoints in FrameworkElementViewManager I can see that the right code is getting hit and the Visibility value gets set to "Collapsed" - but I never see this in XAML) - any idea what's going on?
> > 
> > 
> > if the tree dump says it's not collapsed then it's probably not collapsed : ) (and something else may be re-setting the property?)
> > Do you see the collapsed element in the UI? What does the Live Visual Tree and Property panes say?
> 
> The "collapsed" element is no longer seen in the UI, but the Live Visual Tree and accompanying properties do not seem to change when display:none is toggled. If I undo the code change I added, same thing happens - component goes out of view when display is set to none, but you are still able to tab into those components - when I add the change back in, I'm no longer able to tab into components which are set to display:none, but according to the properties the visibility is still "visible"...

Final comment before merging: Confirmed the elements are getting collapsed through Live Visual Tree and Property panes however the snapshot tests do not reflect this. Might require filing an issue - and if/when this gets fixed this test will fail and will just need updated snapshots to go through. 

Additionally, I had to hard code the width for the view containing 2 Text and 2 Switch components as there was a weird width mismatch from running locally vs in the CI checks. To confirm the display:none property change was not the cause, I made sure update property for display was not getting hit when the prop was not used and commit a2ea3df confirmed the view width issue persists even when display:none is not used.
